File: AEDS-02/Tps/Tp_2026/TP04/Questao07/Questao07.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <string.h>

typedef struct Data
{
    int dia;
    int mes;
    int ano;
} Data;

typedef struct Hora
{
    int hora;
    int minuto;
} Hora;

typedef struct Restaurante
{
    int id;
    char nome[100];
    char cidade[100];
    int capacidade;
    double avaliacao;
    char tipo1[50];
    char tipo2[50];
    char faixaPreco[10];
    Hora horario_abertura;
    Hora horario_fechamento;
    Data data_abertura;
    int aberto;
} Restaurante;

typedef struct Colecao_restaurante
{
    int tamanho;
    Restaurante restaurantes[500];
} Colecao_restaurante;

typedef struct Celula
{
    Restaurante *restaurante;
    struct Celula *prox;
} Celula;

typedef struct No
{
    char letra;
    Celula *primeiro;
    struct No *esq;
    struct No *dir;
} No;

typedef struct ArvoreLista
{
    No *raiz;
    int comparacoes;
} ArvoreLista;

/**
 * @author Gabriel Ferreira Pereira
 * @param s String no formato YYYY-MM-DD
 * @reason Converte uma String para um objeto Data
 * @return objeto Data correspondente
 */
Data parse_data( char *s )
{
    Data data;
    sscanf(s, "%d-%d-%d", &data.ano, &data.mes, &data.dia);
    return data;
}

/**
 * @author Gabriel Ferreira Pereira
 * @param data objeto Data
 * @param saida_linha para armazenar a data formatada
 * @reason Retorna a data no formato DD/MM/YYYY
 */
void formatar_data( Data *data, char *saida_linha )
{
    sprintf(saida_linha, "%02d/%02d/%04d", data->dia, data->mes, data->ano);
}

/**
 * @author Gabriel Ferreira Pereira
 * @param s String no formato hh:mm
 * @reason Converte uma String para um objeto Hora
 * @return objeto Hora correspondente
 */
Hora parse_hora( char *s )
{
    Hora hora;
    sscanf(s, "%d:%d", &hora.hora, &hora.minuto);
    return hora;
}

/**
 * @author Gabriel Ferreira Pereira
 * @param hora objeto Hora
 * @param saida_linha para armazenar a hora formatada
 * @reason Retorna a hora no formato hh:mm
 */
void formatar_hora( Hora *hora, char *saida_linha )
{
    sprintf(saida_linha, "%02d:%02d", hora->hora, hora->minuto);
}

/**
 * @author Gabriel Ferreira Pereira
 * @param linha, pos, campo
 * @reason Le um campo da linha do CSV ate encontrar virgula ou fim de linha
 * @return nova posicao na linha apos o campo lido
 */
int ler_campo( char *linha, int pos, char *campo )
{
    int i = 0;
    
    while ( linha[pos] != ',' && linha[pos] != '\0' && linha[pos] != '\n' )
    {
        campo[i] = linha[pos];
        i++;
        pos++;
    }

    campo[i] = '\0';
    pos++;

    return pos;
}

/**
 * @author Gabriel Ferreira Pereira
 * @param linha String da linha do CSV
 * @reason Le uma linha do CSV e cria um objeto Restaurante
 * @return Restaurante criado
 */
Restaurante ler_restaurante( char *linha )
{
    Restaurante r;
    char campo[100];
    char horario_abertura[10];
    char horario_fechamento[10];
    char data_str[20];
    char aberto_str[10];
    char cozinha[100];
    int pos = 0;
    int j = 0, k = 0;

    pos = ler_campo(linha, pos, campo);
    sscanf(campo, "%d", &r.id);

    pos = ler_campo(linha, pos, r.nome);
    pos = ler_campo(linha, pos, r.cidade);

    pos = ler_campo(linha, pos, campo);
    sscanf(campo, "%d", &r.capacidade);

    pos = ler_campo(linha, pos, campo);
    sscanf(campo, "%lf", &r.avaliacao);

    pos = ler_campo(linha, pos, cozinha);

    while ( cozinha[j] != ';' )
    {
        r.tipo1[k++] = cozinha[j++];
    }

    r.tipo1[k] = '\0';
    j++;
    k = 0;
    
    while ( cozinha[j] != '\0' )
    {
        r.tipo2[k++] = cozinha[j++];
    }
    r.tipo2[k] = '\0';

    pos = ler_campo(linha, pos, r.faixaPreco);
    pos = ler_campo(linha, pos, campo);
    j = 0;
    k = 0;

    while ( campo[j] != '-' )
    {
        horario_abertura[k++] = campo[j++];
    }

    horario_abertura[k] = '\0';
    j++;
    k = 0;

    while ( campo[j] != '\0' )
    {
        horario_fechamento[k++] = campo[j++];
    }

    horario_fechamento[k] = '\0';
    r.horario_abertura = parse_hora(horario_abertura);
    r.horario_fechamento = parse_hora(horario_fechamento);

    pos = ler_campo(linha, pos, data_str);
    r.data_abertura = parse_data(data_str);

    pos = ler_campo(linha, pos, aberto_str);

    if ( aberto_str[0] == 't' )
    {
        r.aberto = 1;
    }

    else
    {
        r.aberto = 0;
    }

    return r;
}

/**
 * @author Gabriel Ferreira Pereira
 * @param r, saida_linha
 * @reason Retorna o restaurante formatado como String
 */
void formatar_restaurante( Restaurante *r, char *saida_linha )
{
    char data_str[20];
    char horario_abertura[10];
    char horario_fechamento[10];
    char aberto_str[10];

    formatar_data(&r->data_abertura, data_str);
    formatar_hora(&r->horario_abertura, horario_abertura);
    formatar_hora(&r->horario_fechamento, horario_fechamento);

    if ( r->aberto == 1 )
    {
        sprintf(aberto_str, "true");
    }

    else
    {
        sprintf(aberto_str, "false");
    }

    sprintf(saida_linha, "[%d ## %s ## %s ## %d ## %.1f ## [%s,%s] ## %s ## %s-%s ## %s ## %s]",
        r->id, r->nome, r->cidade, r->capacidade, r->avaliacao,
        r->tipo1, r->tipo2, r->faixaPreco,
        horario_abertura, horario_fechamento,
        data_str, aberto_str);
}

/**
 * @author Gabriel Ferreira Pereira
 * @param colecao objeto Colecao_restaurante
 * @reason Retorna a quantidade de restaurantes da colecao
 * @return tamanho da colecao
 */
int get_tamanho( Colecao_restaurante *colecao )
{
    return colecao->tamanho;
}

/**
 * @author Gabriel Ferreira Pereira
 * @param colecao, r
 * @reason Adiciona um restaurante na colecao
 */
void adicionar( Colecao_restaurante *colecao, Restaurante r )
{
    colecao->restaurantes[colecao->tamanho] = r;
    colecao->tamanho++;
}

/**
 * @author Gabriel Ferreira Pereira
 * @param colecao objeto Colecao_restaurante
 * @reason Retorna o array de restaurantes da colecao
 * @return ponteiro para o array de restaurantes
 */
Restaurante* get_restaurantes( Colecao_restaurante *colecao )
{
    return colecao->restaurantes;
}

/**
 * @author Gabriel Ferreira Pereira
 * @param colecao objeto Colecao_restaurante
 * @reason Imprime todos os restaurantes da colecao formatados
 */
void imprimir( Colecao_restaurante *colecao )
{
    char saida_linha[500];
    for ( int i = 0; i < colecao->tamanho; i++ )
    {
        formatar_restaurante(&colecao->restaurantes[i], saida_linha);
        printf("%s\n", saida_linha);
    }
}

/**
 * @author Gabriel Ferreira Pereira
 * @reason Le o dataset do arquivo CSV e retorna a colecao de restaurantes
 * @return colecao de restaurantes
 */
Colecao_restaurante ler_csv()
{
    Colecao_restaurante colecao;
    colecao.tamanho = 0;

    FILE *f = fopen("/tmp/restaurantes.csv", "r");

    char linha[500];

    if ( f == NULL )
    {
        return colecao;
    }
 
    if ( fgets(linha, 500, f) == NULL )
    {
        fclose(f);
        return colecao;
    }

    for ( int i = 0; i < 500 && fgets(linha, 500, f) != NULL; i++ )
    {
        int j = 0;

        while ( linha[j] != '\n' && linha[j] != '\r' && linha[j] != '\0' )
        {
            j++;
        }
        linha[j] = '\0';


        Restaurante r = ler_restaurante(linha);
        adicionar(&colecao, r);
    }

    fclose(f);
    return colecao;
}

Celula *nova_celula( Restaurante *restaurante )
{
    Celula *nova = (Celula*) malloc(sizeof(Celula));
    nova->restaurante = restaurante;
    nova->prox = NULL;
    return nova;
}

No *novo_no( char letra )
{
    No *novo = (No*) malloc(sizeof(No));
    novo->letra = letra;
    novo->primeiro = nova_celula(NULL);
    novo->esq = NULL;
    novo->dir = NULL;
    return novo;
}

void iniciar_arvore_lista( ArvoreLista *arvore )
{
    arvore->raiz = NULL;
    arvore->comparacoes = 0;
}

void inserir_na_lista( No *no, Restaurante *restaurante, int *comparacoes )
{
    Celula *ant = no->primeiro;
    Celula *i = ant->prox;

    while ( i != NULL && strcmp(i->restaurante->nome, restaurante->nome) < 0 )
    {
        (*comparacoes)++;
        ant = i;
        i = i->prox;
    }

    if ( i != NULL )
    {
        (*comparacoes)++;
    }

    Celula *nova = nova_celula(restaurante);
    nova->prox = i;
    ant->prox = nova;
}

No *inserir_no( No *i, Restaurante *restaurante, int *comparacoes )
{
    char letra = restaurante->nome[0];

    if ( i == NULL )
    {
        i = novo_no(letra);
        inserir_na_lista(i, restaurante, comparacoes);
    }

    else if ( letra < i->letra )
    {
        (*comparacoes)++;
        i->esq = inserir_no(i->esq, restaurante, comparacoes);
    }

    else if ( letra > i->letra )
    {
        (*comparacoes)++;
        i->dir = inserir_no(i->dir, restaurante, comparacoes);
    }

    else
    {
        (*comparacoes)++;
        inserir_na_lista(i, restaurante, comparacoes);
    }

    return i;
}

void inserir_arvore_lista( ArvoreLista *arvore, Restaurante *restaurante )
{
    arvore->raiz = inserir_no(arvore->raiz, restaurante, &arvore->comparacoes);
}

int pesquisar_lista( No *no, char *nome, Restaurante *encontrado, int *comparacoes )
{
    int resp = 0;
    Celula *i = no->primeiro->prox;

    while ( i != NULL && strcmp(i->restaurante->nome, nome) < 0 )
    {
        (*comparacoes)++;
        printf("%s ", i->restaurante->nome);
        i = i->prox;
    }

    if ( i != NULL )
    {
        (*comparacoes)++;
        if ( strcmp(i->restaurante->nome, nome) == 0 )
        {
            *encontrado = *(i->restaurante);
            resp = 1;
        }
    }

    return resp;
}

int pesquisar_no( No *i, char *nome, Restaurante *encontrado, int *comparacoes )
{
    int resp = 0;
    char letra = nome[0];

    if ( i != NULL )
    {
        (*comparacoes)++;

        if ( letra < i->letra )
        {
            printf("ESQ ");
            resp = pesquisar_no(i->esq, nome, encontrado, comparacoes);
        }

        else if ( letra > i->letra )
        {
            printf("DIR ");
            resp = pesquisar_no(i->dir, nome, encontrado, comparacoes);
        }

        else
        {
            resp = pesquisar_lista(i, nome, encontrado, comparacoes);
        }
    }

    return resp;
}

int pesquisar_arvore_lista( ArvoreLista *arvore, char *nome, Restaurante *encontrado )
{
    int resp = 0;
    printf("RAIZ ");
    resp = pesquisar_no(arvore->raiz, nome, encontrado, &arvore->comparacoes);
    return resp;
}

/**
 * @author Gabriel Ferreira Pereira
 * @reason Metodo principal que busca e formata o restaurante com o ID fornecido
 *         e exibe na tela a lista de restaurantes selecionados
 */
int main()
{
    Colecao_restaurante colecao = ler_csv();
    Restaurante *restaurantes = get_restaurantes(&colecao);
    ArvoreLista arvore;
    Restaurante encontrado;
    char saida_linha[500];
    int id = 0;

    iniciar_arvore_lista(&arvore);

    while ( scanf("%d", &id) && id != -1 )
    {
        for ( int i = 0; i < get_tamanho(&colecao); i++ )
        {
            if ( restaurantes[i].id == id )
            {
                inserir_arvore_lista(&arvore, &restaurantes[i]);
                i = get_tamanho(&colecao);
            }
        }
    }

    char nome[100];
    fgets(nome, 100, stdin);

    clock_t inicio = clock();

    while ( fgets(nome, 100, stdin) != NULL )
    {
        int j = 0;

        while ( nome[j] != '\n' && nome[j] != '\r' && nome[j] != '\0' )
        {
            j++;
        }

        nome[j] = '\0';

        if ( strcmp(nome, "FIM") != 0 )
        {
            int resp = pesquisar_arvore_lista(&arvore, nome, &encontrado);

            if ( resp == 0 )
            {
                printf("NAO\n");
            }
            else
            {
                formatar_restaurante(&encontrado, saida_linha);
                printf("SIM %s\n", saida_linha);
            }
        }
    }

    clock_t fim = clock();
    double total = (double) (fim - inicio) / CLOCKS_PER_SEC;

    FILE *log = fopen("842527_hibrida_arvore_lista.txt", "w");

    fprintf(log, "Comparacoes: %d\n", arvore.comparacoes);
    fprintf(log, "Total: %f\n", total);
    fclose(log);
    
}

File: AEDS-02/Tps/Tp_2026/TP02/Questao07/Questao07.java
import java.io.File;
import java.io.PrintWriter;
import java.util.*;

class Data
{
    private int ano;
    private int mes;
    private int dia;
    
    /**
     * @author Gabriel Ferreira Pereira
     * @param ano mes e dia da data
     * @reason Construtor da classe Data
     */
    public Data ( int ano, int mes, int dia ) 
    {
        this.ano = ano;
        this.mes = mes;
        this.dia = dia;    
    }

    /**
     * @author Gabriel Ferreira Pereira
     * @param s String no formato YYYY-MM-DD
     * @reason Converte uma String para um objeto Data
     * @return objeto Data correspondente
     */
    public static Data parseData( String s )
    {
        Scanner sc = new Scanner(s);
        sc.useDelimiter("-");
        int ano = sc.nextInt();
        int mes = sc.nextInt();
        int dia = sc.nextInt();

        Data data = new Data ( ano, mes, dia );
        
        return data;
    }
    
    /**
     * @author Gabriel Ferreira Pereira
     * @reason Retorna a data no formato DD/MM/YYYY
     * @return String com a data formatada
     */
    public String formatarData ()
    {
        String s = String.format("%02d/%02d/%04d", dia,mes,ano);
        return s;
    }
}


class Hora
{
    private int hora;
    private int minuto;

    /**
     * @author Gabriel Ferreira Pereira
     * @param hora e minuto da hora
     * @reason Construtor da classe Hora
     */
    public Hora ( int hora, int minuto )
    {
        this.hora = hora;
        this.minuto = minuto;
    }

    /**
     * @author Gabriel Ferreira Pereira
     * @param s String no formato HH:mm
     * @reason Converte uma String para um objeto Hora
     * @return objeto Hora correspondente
     */
    public static Hora parseHora ( String s )
    {
        Scanner sc = new Scanner(s);
        sc.useDelimiter(":");
        int hora1 = sc.nextInt();
        int minuto = sc.nextInt();

        Hora hora = new Hora(hora1, minuto);

        return hora;
    }
    
    /**
     * @author Gabriel Ferreira Pereira
     * @reason Retorna a hora no formato HH:mm
     * @return String com a hora formatada
     */
    public String formatarHora ()
    {
        String s = String.format("%02d:%02d", hora, minuto);
        return s;
    }
}


class Restaurante
{
    private int id;
    private String nome;
    private String cidade;
    private int capacidade;
    private double avaliacao;
    private String[] tiposCozinha;
    private String faixaPreco;
    private Hora horarioAbertura;
    private Hora horarioFechamento;
    private Data dataAbertura;
    private boolean aberto;
    

    /**
     * @author Gabriel Ferreira Pereira
     * @param id, nome, cidade, capacidade, avaliacao, tiposCozinha, faixaPreco, horarioAbertura, 
     *          horarioFechamento, dataAbertura, aberto do restaurante
     * @reason Construtor da classe Restaurante
     */
    public Restaurante( int id, String nome, String cidade, int capacidade, double avaliacao, 
        String[] tiposCozinha, String faixaPreco, Hora horarioAbertura, Hora horarioFechamento, Data dataAbertura, boolean aberto ) 
    {
        this.id = id;
        this.nome = nome;
        this.cidade = cidade;
        this.capacidade = capacidade;
        this.avaliacao = avaliacao;
        this.tiposCozinha = tiposCozinha;
        this.faixaPreco = faixaPreco;
        this.horarioAbertura = horarioAbertura;
        this.horarioFechamento = horarioFechamento;
        this.dataAbertura = dataAbertura;
        this.aberto = aberto;
    }

    // Getters

    public int getID ()
    {
        return this.id;
    }

    public String getNome ()
    {
        return this.nome;
    }

    public String getCidade ()
    {
        return this.cidade;
    }

    public int getCapacidade ()
    {
        return this.capacidade;
    }

    public double getAvaliacao ()
    {
        return this.avaliacao;
    }

    public String[] getTiposCozinha ()
    {
        return this.tiposCozinha;
    }

    public String getFaixaPreco ()
    {
        return this.faixaPreco;
    }

    public Hora getHorarioAbertura ()
    {
        return this.horarioAbertura;
    }

    public Hora getHorarioFechamento ()
    {
        return this.horarioFechamento;
    }

    public Data getDataAbertura ()
    {
        return this.dataAbertura;
    }

    public boolean getAberto ()
    {
        return this.aberto;
    }
    
    /**
     * @author Gabriel Ferreira Pereira
     * @param String s
     * @reason Converte uma String para double manualmente
     * @return numero real
     */
    public static double parseDouble( String s )
    {
        int inteiro = 0;
        int decimal = 0;
        boolean ponto = false;

        for ( int i = 0; i < s.length(); i++ )
        {
            char c = s.charAt(i);
            
            if (c == '.')
            {
                ponto = true;
            }
            
            else if ( ponto == false) 
            {
                inteiro = inteiro * 10 + (c - '0');
            }

            else
            {
                decimal = decimal * 10 + (c - '0');
            }
        }

        return inteiro + decimal / 10.0;
    }

    /**
     * @author Gabriel Ferreira Pereira
     * @param path caminho do arquivo CSV
     * @reason Le uma linha do CSV e cria um objeto Restaurante
     * @return Restaurante criado
     */
    public static Restaurante ler( String linha ) throws Exception
    {
        Scanner sc = new Scanner(linha);
        sc.useDelimiter(",");

        int id = sc.nextInt();
        String nome = sc.next();
        String cidade = sc.next();
        int capacidade = sc.nextInt();
        double avaliacao = parseDouble(sc.next());

        String cozinhaStr = sc.next();
        Scanner scCozinha = new Scanner(cozinhaStr);
        scCozinha.useDelimiter(";");
        String tipo1 = scCozinha.next();
        String tipo2 = scCozinha.next();
        scCozinha.close();
        String[] tiposCozinha = new String[]{tipo1, tipo2};

        String faixaPreco = sc.next();

        String horario = sc.next();
        Scanner scHorario = new Scanner(horario);

        scHorario.useDelimiter("-");

        Hora horarioAbertura   = Hora.parseHora(scHorario.next());
        Hora horarioFechamento = Hora.parseHora(scHorario.next());
        scHorario.close();

        Data dataAbertura = Data.parseData(sc.next());

        String abertoStr = sc.next();
        boolean aberto;

        if (abertoStr.charAt(0) == 't')
        {
            aberto = true;
        }

        else
        {
            aberto = false;
        }

        Restaurante restaurante = new Restaurante(id, nome, cidade, capacidade, avaliacao, tiposCozinha, faixaPreco, horarioAbertura, horarioFechamento, dataAbertura, aberto);
        return restaurante;    
    }

    /**
     * @author Gabriel Ferreira Pereira
     * @reason Retorna o restaurante formatado como String
     * @return String com o restaurante formatado
     */
    public String formatar()
    {
        String tipos = "[" + tiposCozinha[0] + "," + tiposCozinha[1] + "]";
        String horario = horarioAbertura.formatarHora() + "-" + horarioFechamento.formatarHora();
        String data = dataAbertura.formatarData();
    
        return "[" + id + " ## " + nome + " ## " + cidade + " ## " + capacidade + " ## " + 
            avaliacao + " ## " + tipos + " ## " + faixaPreco + " ## " + 
            horario + " ## " + data + " ## " + aberto + "]";
    }

}

class ColecaoRestaurante 
{

    private int tamanho;
    private Restaurante[] restaurantes;

    public ColecaoRestaurante( int tamanho, Restaurante[] restaurantes ) 
    {
        this.tamanho = tamanho;
        this.restaurantes = restaurantes;
    }

    /**
     * @author Gabriel Ferreira Pereira
     * @reason Retorna a quantidade de restaurantes da colecao
     * @return this.tamanho
     */
    public int getTamanho() 
    {
        return this.tamanho;
    }

    /**
     * @author Gabriel Ferreira Pereira
     * @param Restaurante r
     * @reason Adiciona tamanho em ColecaoRestaurante
     */
    public void adicionar( Restaurante r )
    {
        this.restaurantes[this.tamanho] = r;
        this.tamanho++;
    }
    /**
     * @author Gabriel Ferreira Pereira
     * @reason Pega os restaurantes da colecao
     * @return this.restaurantes
     */
    public Restaurante[] getRestaurantes() 
    {
        return this.restaurantes;
    }

    /**
     * @author Gabriel Ferreira Pereira
     * @reason Imprime todos os restaurantes da colecao
     */
    public void imprimir() 
    {
        for ( int i = 0; i < this.tamanho; i++ ) 
        {
            System.out.println(this.restaurantes[i].formatar());
        }
    }

    /**
     * @author Gabriel Ferreira Pereira
     * @reason Le o dataset do arquivo CSV usando Restaurante.ler() e retorna a colecao
     * @return colecao
     */
    public static ColecaoRestaurante lerCsv() throws Exception
    {
        ColecaoRestaurante colecao = new ColecaoRestaurante(0, new Restaurante[550]);
        Scanner sc = new Scanner(new File("/tmp/restaurantes.csv"));
        
        sc.nextLine();

        sc.useDelimiter(",|\n");

        for ( int i = 0; i < 500; i++ )
        {
            colecao.adicionar(Restaurante.ler(sc.nextLine()));
        }

        return colecao;
    }
}


class Questao07 
{
    public static void mergesort( Restaurante[] restaurantes, int esq, int dir, int[] contadores )
    {
        if ( esq < dir )
        {
            int meio = ( esq + dir ) / 2;
            mergesort( restaurantes, esq, meio, contadores );
            mergesort( restaurantes, meio + 1, dir, contadores );
            intercalar( restaurantes, esq, meio, dir, contadores );
        }
    }

    public static void intercalar( Restaurante[] restaurantes, int esq, int meio, int dir, int[] contadores )
    {
        int n1, n2, i, j, k;
        n1 = meio - esq + 1;
        n2 = dir - meio;

        Restaurante[] a1 = new Restaurante[n1];
        Restaurante[] a2 = new Restaurante[n2];

        for ( i = 0; i < n1; i++ )
        {
            a1[i] = restaurantes[esq + i];
        }

        for ( j = 0; j < n2; j++ )
        {
            a2[j] = restaurantes[meio + j + 1];
        }

        i = 0;
        j = 0;
        k = esq;

        // Intercala os dois subarrays comparando cidade e desempatando por nome
        
        while ( i < n1 && j < n2 )
        {
            contadores[0]++; // avanca o movimentacoes

            int tmp = a1[i].getCidade().compareTo( a2[j].getCidade() );

            // Se as cidades forem iguais, desempata pelo nome
            if ( tmp == 0 )
            {
                contadores[0]++; // avanca o movimentacoes
                tmp = a1[i].getNome().compareTo( a2[j].getNome() );
            }

            // Coloca o menor elemento no array original
            if ( tmp <= 0 )
            {
                restaurantes[k++] = a1[i++];
            }
            else
            {
                restaurantes[k++] = a2[j++];
            }

            contadores[1]++; // avanca o comparacoes
        }

        // Copia o restante de a1 caso a2 tenha acabado primeiro
        while ( i < n1 )
        {
            restaurantes[k++] = a1[i++];
            contadores[1]++; // avanca o comparcoes 
        }

        // Copia o restante de a2 caso a1 tenha acabado primeiro
        while ( j < n2 )
        {
            restaurantes[k++] = a2[j++];
            contadores[1]++; // avanca o comparacoes
        }
    }

    /**
     * @author Gabriel Ferreira Pereira
     * @reason Metodo principal que busca e formata o restaurante com o ID fornecido
     *         e exibe na tela a lista de restaurantes selecionados
     */
    public static void main(String[] args) throws Exception
    {
        Scanner sc = new Scanner(System.in);
        
        ColecaoRestaurante colecao = ColecaoRestaurante.lerCsv();
        Restaurante[] restaurantes = colecao.getRestaurantes();
        Restaurante[] selecionados = new Restaurante[500];

        int tamanho = 0;
        
        int id = 0;

        long inicio, fim;

        while ( ( id = sc.nextInt() ) != -1 )
        {
            for (int i = 0; i < colecao.getTamanho(); i++)
            {
                if (restaurantes[i].getID() == id)
                {
                    selecionados[tamanho] = restaurantes[i];
                    tamanho++;
                }
            }
        }
  
        // Execucao do algoritmo de ordenacao
        int[] contadores = new int[]{0, 0};
        
        inicio = System.currentTimeMillis();

        mergesort( selecionados, 0, tamanho - 1, contadores );

        fim = System.currentTimeMillis();

        int comparacoes = contadores[0];
        int movimentacoes = contadores[1];

        boolean ordenado = true;
        
        for (int i = 0; i < tamanho - 1; i++)
        {
            // Compara a cidade do elemento atual com a cidade do proximo elemento
            // Se a cidade atual for maior que a proxima, o array nao esta ordenado
            
            if (selecionados[i].getCidade().compareTo(selecionados[i + 1].getCidade()) > 0)
            {
                ordenado = false;
            }
        }
        
        //Mostrar o conjunto ordenado, tempo de execucao e status da ordenacao
		//algoritmo.mostrar();
        // Salvar tempo e status em arquivo
        PrintWriter log = new PrintWriter("842527_mergesort.txt");
        log.println("Tempo para ordenar: " + (fim - inicio) / 1000.0 + " s.");
        log.println("isOrdenado: " + ordenado);
        log.println("Comparacoes: " + comparacoes);
        log.println("Movimentacoes: " + movimentacoes);
        log.close();


        for (int i = 0; i < tamanho; i++)
        {
            System.out.println(selecionados[i].formatar());
        }
    }
}
